src/compas_rhino/geometry/trimesh/curvature.py

```diff
@@ -0,0 +1,331 @@
+            normals = dict(zip(orientation, [faces_normal[faces[0]], faces_normal[faces[1]]]))
+            e = mesh.TopologyEdges.EdgeLine(edge).Direction
+            e.Unitize()
+            n1 = normals[True]
+            n2 = normals[False]
+            if not direction:
+                e.Reverse()
+                n1, n2 = n2, n1
+            # (3.3) calculate dihedral angle
+            angle = dihedral_angle(e, n1, n2)
+            x.append(l_ij * angle)
+        H.append(1/4 * sum(x))
+
+    # (4) Output
+    return H
+
+
+@pluggable(category="trimesh", requires=['Rhino'])
+def trimesh_principal_curvature(M):
+    """Compute the principal curvature of a triangle mesh.
+    Parameters
+    ----------
+    M : (list, list)
+        A mesh represented by a list of vertices and a list of faces.
+    Returns
+    -------
+    Curvature : (k1, k2)
+        k1_list, the max curvature per vertex.
+        k2_list, the min curvature per vertex.
+
+    Notes
+    -----
+    Description: The discrete principal curvature is computed by mean curvature, Gaussian curvature, and vertex area.
+
+    Notation Convention:
+    :math:`\kappa_1, \kappa_2` - The max principal curvature and the min principal curvature
+    :math:`H_i` - the discrete mean curvature of vertex i
+    :math:`K_i` - the discrete Gaussian curvature of vertex i
+    :math:`A_i` - the vertex area of vertex i
+
+    Formula:
+    .. math::
+
+        \frac{H_i}{A_i}\pm\sqrt{( \,\frac{H_i}{A_i})\,^2-\frac{K_i}{A_i}}
+
+    Examples
+    --------
+    Make a mesh from scratch
+    >>> from compas.geometry import Sphere
+    >>> sphere = Sphere([1, 1, 1], 1)
+    >>> sphere = Mesh.from_shape(sphere, u=30, v=30)
+    >>> sphere.quads_to_triangles()
+    >>> M = sphere.to_vertices_and_faces()
+
+    Compute the discrete principal curvature
+
+    >>> H = trimesh_principal_curvature(M)
+
+    References
+    ----------
+    .. [1] Formula of Discrete Principal Curvature available at Keenan Crane's lecture, 03:16-07:11, at https://youtu.be/sokeN5VxBB8
+
+    """
+    # (1) see if input is already Rhino.Geometry.Mesh
+    mesh = Rhino.Geometry.Mesh()
+    if type(M) != Rhino.Geometry.Mesh:
+        for vertices, faces in M:
+            for x, y, z in vertices:
+                mesh.Vertices.Add(x, y, z)
+            for face in faces:
+                mesh.Faces.AddFace(*face)
+    else:
+        mesh = M
+
+    # (2) Prepare ingredient and return list
+    k1 = []
+    k2 = []
+    A = trimesh_barycentric_area(mesh)
+    H = trimesh_mean_curvature(mesh)
+    K = trimesh_gaussian_curvature(mesh)
+
+    # (3) Main - loop over all vertices
+    for i in range(mesh.Vertices.Count):
+        if(A[i] == 0):
+            k1.append(0)
+            k2.append(0)
+            continue
+        H_ = H[i] / A[i]
+        K_ = K[i] / A[i]
+
+        k1.append(H_ + sqrt(max(0, H_ * H_ - K_)))
+        k2.append(H_ - sqrt(max(0, H_ * H_ - K_)))
```

Comment:
`sqrt` is expensive, and to do this twice for every vertex in a mesh might be slower than necessary.  
```suggestion
        discriminant = sqrt(max(0, H_ * H_ - K_))
        k1.append(H_ + discriminant)
        k2.append(H_ - discriminant)
```